### PR TITLE
More-correct typings for chaining in underscore

### PR DESCRIPTION
**Pull request by jacobbaskin**
In particular, the wrapped collection type in `_Chain` was often lost, which didn't cause problems until returning a value. Anything of type `_Chain<T>` is typed to return a single `T` from its `.value()` when generally what is actually returned is a `T[]` or `Dictionary<T>`.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: [see main underscore docs](http://underscorejs.org) for context 
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

---
This change is [Reviewable](https://reviewable.io/reviews/definitelytyped/definitelytyped/36838)

### CODE REVIEW

File: types/underscore/index.d.ts [hunk 1]
```diff
@@ -5186,18 +5186,18 @@ declare module _ {
         * Wrapped type `any[]`.
         * @see _.filter
         **/
-        filter(iterator: _.ListIterator<T, boolean>, context?: any): _Chain<T>;
+        filter(iterator: _.ListIterator<T, boolean>, context?: any): _Chain<T, T[]>;
 
         /**
-        * @see _.filter
+        * @see _.select
```
Comment (regevbr) [inline types/underscore/index.d.ts]: actually the original one is better as it is easier to find in the unserscore website using the formal name and not the alias

File: types/underscore/index.d.ts [hunk 2]
```diff
@@ -5255,7 +5255,7 @@ declare module _ {
         * Wrapped type `any[]`.
         * @see _.invoke
         **/
-        invoke(methodName: string, ...args: any[]): _Chain<T>;
+        invoke(methodName: string, ...args: any[]): _Chain<T, V>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: This is actually unknown as you don't know the return value of the method

File: types/underscore/index.d.ts [hunk 3]
```diff
@@ -5327,48 +5327,48 @@ declare module _ {
...
         * @see _.sample
         **/
-        sample<T>(n: number): _Chain<T>;
+        sample<T>(n: number): _Chain<T, T[]>;
 
         /**
         * @see _.sample
         **/
-        sample<T>(): _Chain<T>;
+        sample<T>(): _Chain<T, T[]>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: this actually retuns a single value - please use _ChainSingle

File: types/underscore/index.d.ts [hunk 4]
```diff
@@ -5691,60 +5691,60 @@ declare module _ {
...
         **/
-        allKeys(): _Chain<string>;
+        allKeys(): _Chain<string, string[]>;
 
         /**
         * Wrapped type `object`.
         * @see _.values
         **/
-        values(): _Chain<any>;
+        values(): _Chain<any, any[]>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: I believe this should be <T, T[]>

File: types/underscore/index.d.ts [hunk 5]
```diff
@@ -5691,60 +5691,60 @@ declare module _ {
...
 
         /**
         * Wrapped type `object`.
         * @see _.mapObject
         **/
-        mapObject(fn: _.ListIterator<T, any>): _Chain<T>;
+        mapObject<TResult>(fn: _.ListIterator<T, TResult>): _Chain<TResult, any>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: Shoulbe be  _Chain<TResult, Dictionary<TResult>>;

File: types/underscore/index.d.ts [hunk 6]
```diff
@@ -5691,60 +5691,60 @@ declare module _ {
...
 
         /**
         * Wrapped type `object`.
         * @see _.pairs
         **/
-        pairs(): _Chain<[string, any]>;
+        pairs(): _Chain<[string, any], [string, any][]>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: _Chain<[string, T], [string, T][]>;

**Comment by regevbr** [comment]
@jacobbaskin thanks for the PR - I added some comments for you to review.
We are in the process of making a proper fix for the typings - please see #36532
@jgonggrijp would you mind to go over the PR as well? - it covers a lot of ground and I want to make sure I didn't miss anything

File: types/underscore/index.d.ts (hunk 1 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Good point, reverted.

File: types/underscore/index.d.ts (hunk 2 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Changed to _Chain<any, any[]>

File: types/underscore/index.d.ts (hunk 3 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done.

File: types/underscore/index.d.ts (hunk 4 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done.

File: types/underscore/index.d.ts (hunk 5 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done.

File: types/underscore/index.d.ts (hunk 6 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done.

File: types/underscore/index.d.ts [hunk 7]
```diff
@@ -5357,12 +5357,12 @@ declare module _ {
         * Wrapped type `any[]`.
         * @see _.sample
         **/
-        sample<T>(n: number): _Chain<T, T[]>;
+        sample<T>(n: number): _ChainSingle<T>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: This actually does return an array (unlike  sample<T>())

**Review by regevbr — CHANGES_REQUESTED** [review]

File: types/underscore/index.d.ts (hunk 7 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done. I also added a test --- the <T> here was breaking things.

File: types/underscore/index.d.ts [hunk 8]
```diff
@@ -5758,177 +5758,177 @@ declare module _ {
         * Wrapped type `object`.
         * @see _.omit
         **/
-        omit(...keys: string[]): _Chain<T>;
-        omit(keys: string[]): _Chain<T>;
-        omit(iteratee: Function): _Chain<T>;
+        omit(...keys: string[]): _Chain<T, V>;
+        omit(keys: string[]): _Chain<T, V>;
```
Comment (jgonggrijp) [inline types/underscore/index.d.ts]: I wonder whether it would be possible to do this more accurately, something like this:

```ts
omit<U extends string>(keys: U[]): _Chain<Omit<T, V[U]>, Omit<V, >>;
```

But this is probably beyond the scope of this pull request and it might require a TS upgrade, so definitely not a condition for acceptance.

File: types/underscore/index.d.ts [hunk 9]
```diff
@@ -5691,60 +5691,60 @@ declare module _ {
...
         **/
-        values(): _Chain<any>;
+        values(): _Chain<T, T[]>;
 
         /**
         * Wrapped type `object`.
         * @see _.mapObject
         **/
-        mapObject(fn: _.ListIterator<T, any>): _Chain<T>;
+        mapObject<TResult>(fn: _.ListIterator<T, TResult>): _Chain<TResult, Dictionary<TResult>>;
```
[the remark this answers is not in the document]
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done.

File: types/underscore/index.d.ts (hunk 8 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Yeah if @regevbr is contemplating a complete refactor anyway, I would rather hold off.

File: types/underscore/index.d.ts [hunk 10]
```diff
@@ -5938,78 +5938,78 @@ declare module _ {
...
         **/
-        unescape(): _Chain<T>;
+        unescape(): _ChainSingle<string>;
 
         /**
         * Wrapped type `object`.
         * @see _.result
         **/
-        result(property: string, defaultValue?: any): _Chain<T>;
+        result(property: string, defaultValue?: any): _ChainSingle<any>;
```
Comment (jgonggrijp) [inline types/underscore/index.d.ts]: Maybe possible to be more accurate:

```suggestion
        result<K in keyof V, D>(property: K, defaultValue?: D): _ChainSingle<
            V[K] extends () => infer R ? R | D : V[K] | D
        >;
        result<K not in keyof V, D = undefined>(property: K, defaultValue?: D): _ChainSingle<D>;
```

File: types/underscore/underscore-tests.ts [hunk 11]
```diff
@@ -571,6 +571,31 @@ function chain_tests() {
...
+    .value()
+    .length;
+
+  var groupBy = _.chain(['a', 'aa', 'b', 'bb'])
+    .groupBy(v => v[0])
+    .value()[0].length;
+
+  var sampleOne = _.chain([1, 2, 3]).sample().value() * 2;
+  var sampleMore = _.chain([1, 2, 3]).sample(2).value().length;
+
```
Comment (jgonggrijp) [inline types/underscore/underscore-tests.ts]: These tests are good, but they don't cover all the methods you have fixed. Please try to come up with a test for each method that you have edited, preferably one that would have failed before your edits.

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
I made a bunch of code suggestions. Disclaimer: I didn't test any of them, and I might be wrong about some of them. Please review carefully.

The only strong condition for acceptance is the addition of more tests, i.e., my last comment.

Overall it looks like a big improvement to me already.

**Comment by jgonggrijp** [comment]
Also, one question @regevbr @jacobbaskin : this PR introduces a lot of return type signatures of `_Chain<T, T[]>`. Is `T[]` really appropriate or should it rather be `List<T>` or even `Array<T>`?

File: types/underscore/index.d.ts (hunk 8 above)
Comment (regevbr) [reply types/underscore/index.d.ts]: indeed we should hold off on that one

File: types/underscore/underscore-tests.ts (hunk 11 above)
Comment (regevbr) [reply types/underscore/underscore-tests.ts]: I agree!

**Comment by regevbr** [comment]
> Also, one question @regevbr @jacobbaskin : this PR introduces a lot of return type signatures of `_Chain<T, T[]>`. Is `T[]` really appropriate or should it rather be `List<T>` or even `Array<T>`?

Well I'm still contemplating on the matter in the work I'm doing. I think it should stay T[] as we know for sure that underscore will return a "real" array. The `List` interface was created (I believe) to help pass along array like objects.

File: types/underscore/index.d.ts (hunk 10 above)
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Doing this would have to involve a bigger refactoring, since right now _ChainSingle is also broken in various ways.

File: types/underscore/index.d.ts [hunk 12]
```diff
@@ -5758,177 +5758,177 @@ declare module _ {
...
+        omit(...keys: string[]): _Chain<T, V>;
+        omit(keys: string[]): _Chain<T, V>;
+        omit(iteratee: Function): _Chain<T, V>;
 
         /**
         * Wrapped type `object`.
         * @see _.defaults
         **/
-        defaults(...defaults: any[]): _Chain<T>;
+        defaults(...defaults: any[]): _Chain<T, V>;
```
[the remark this answers is not in the document]
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done

File: types/underscore/index.d.ts [hunk 13]
```diff
@@ -5758,177 +5758,177 @@ declare module _ {
...
 
         /**
          * Wrapped type `any`.
          * @see _.create
          **/
-        create(props?: object): _Chain<T>;
+        create(props?: object): _Chain<T, V>;
```
[the remark this answers is not in the document]
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: Done.

File: types/underscore/index.d.ts [hunk 14]
```diff
@@ -5758,177 +5758,177 @@ declare module _ {
...
         **/
-        isMatch(): _Chain<T>;
+        isMatch(properties: Partial<V>): _ChainSingle<boolean>;
 
         /**
         * Wrapped type `object`.
         * @see _.isElement
         **/
-        isElement(): _Chain<T>;
+        isElement(): _ChainSingle<boolean>;
```
[the remark this answers is not in the document]
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: I am not having success getting this to compile. I don't think you can use a template parameter in a type guard, nor can you use a type guard as a template parameter itself. Am I missing something?

**Comment by jacobbaskin** [comment]
Added tests for everything and responded to the comments.

File: types/underscore/index.d.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (jgonggrijp) [reply types/underscore/index.d.ts]: Nah, I might just have been too optimistic about the possibilities.

File: types/underscore/index.d.ts (hunk 10 above)
Comment (regevbr) [reply types/underscore/index.d.ts]: Let leave that one be for now

**Review by regevbr — APPROVED** [review]

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
I'm quite happy already. Just a few details, including some questions.

File: types/underscore/underscore-tests.ts [hunk 15]
```diff
@@ -586,16 +586,159 @@ function chain_tests() {
...
+  interface bindAll {
+    a: () => number;
+    b: (a: number) => string;
+  }
+
+  var didBindAll: bindAll = _.chain({a: () => 5, b: (a: number) => a.toString()})
+    .bindAll('a', 'b')
+    .value();
+
+  var partial: Function = _.chain((a: number, b: number) => b - a)
```
[the remark this answers is not in the document]
Comment (regevbr) [reply types/underscore/underscore-tests.ts]: I will keep it in mind indeed

File: types/underscore/underscore-tests.ts [hunk 16]
```diff
@@ -586,16 +586,159 @@ function chain_tests() {
...
+  _.chain(alert).defer('hello');
+  _.chain(alert).delay(5, 'hello');
+  var throttle: (s: string) => void = _.chain(alert).throttle(5).value();
+  var debounce: (s: string) => void = _.chain(alert).debounce(5).value();
+  var once: (s: string) => void = _.chain((s: string) => console.log(s, s)).once().value();
+  var after: (s: string) => void = _.chain((s: string) => console.log(s, s)).after(1).value();
+  var before: (s: string) => void = _.chain((s: string) => console.log(s, s)).before(2).value();
+  var restArgs: Function = _.chain((a: number, b: string[]) => console.log(a, b))
+    .restArgs(1)
+    .value();
```
[the remark this answers is not in the document]
Comment (regevbr) [reply types/underscore/underscore-tests.ts]: I will need to take care of it... But do note that there is no more `restArgs` methods, but `restArguments` https://underscorejs.org/#restArguments

File: types/underscore/index.d.ts [hunk 17]
```diff
@@ -5651,31 +5651,31 @@ declare module _ {
...
          **/
-        restArgs(startIndex?: number): _Chain<T, V>;
+        restArgs(starIndex?: number): _ChainSingle<Function>;
 
         /**
         * Wrapped type `number`.
         * @see _.after
         **/
-        after(func: Function): _Chain<T, V>;
+        after(times: number): _Chain<T, V>;
```
[the remark this answers is not in the document]
Comment (jacobbaskin) [reply types/underscore/index.d.ts]: It is indeed -- fixed (I templated so the type is a bit better)

File: types/underscore/index.d.ts [hunk 18]
```diff
@@ -5721,7 +5721,7 @@ declare module _ {
         * Wrapped type `object`.
         * @see _.invert
         **/
-        invert(): _Chain<string, Dictionary<string>>;
+        invert(): _Chain<any, Dictionary<any>>;
```
Comment (regevbr) [inline types/underscore/index.d.ts]: why did you change that? according to docs the original deceleration is good https://underscorejs.org/#invert

File: types/underscore/underscore-tests.ts [hunk 19]
```diff
@@ -586,16 +586,159 @@ function chain_tests() {
...
+    .negate()
+    .value()(true, false);
+  var compose: (v: string) => string = _.chain(hello).compose(h => h + '!').value()
+  var keys: string[] = _.chain({a: 1, b: 2}).keys().value();
+  var values: number[] = _.chain({a: 1, b: 2}).values().value();
+  var mappedObj: {[key: string]: string} = _.chain({a: 1, b: 2})
+    .mapObject(v => v.toString())
+    .value();
+  var paired: Array<[string, number]> = _.chain({a: 1, b: 2}).pairs().value();
+  var inverted: {[key: string]: string} = _.chain({a: 'b', b: 'a'}).invert().value();
```
[the remark this answers is not in the document]
Comment (regevbr) [reply types/underscore/underscore-tests.ts]: This is actually not true - actual return value is 
```js
console.log(
  _.chain(['a', 'b', 'c'])
    .invert()
    .value()
);
// { a: '0', b: '1', c: '2' }
```

**Review by regevbr — CHANGES_REQUESTED** [review]

File: types/underscore/underscore-tests.ts [hunk 20]
```diff
@@ -685,6 +686,7 @@ function chain_tests() {
     .value();
   var paired: Array<[string, number]> = _.chain({a: 1, b: 2}).pairs().value();
   var inverted: {[key: string]: string} = _.chain({a: 'b', b: 'a'}).invert().value();
+  var invertedArray: {[key: string]: number} = _.chain(['a', 'b', 'c']).invert().value();
```
Comment (jgonggrijp) [inline types/underscore/underscore-tests.ts]: This is probably the reason why you made `_.invert` `_Chain<any, Dictionary<any>>`... my bad.

**Review by jgonggrijp — CHANGES_REQUESTED** [review]
I think you'll be done when you address these last few comments. Sorry for pointing you in the wrong direction with `_.invert`.

**Comment by jgonggrijp** [comment]
Please keep this PR open; I believe it is almost ready to be merged.

**Comment by jacobbaskin** [comment]
Yes sorry I will respond this afternoon, have been busy with other things.

**Comment by kjleitz** [comment]
@jacobbaskin @jgonggrijp @regevbr Anybody intend to open up a new PR with these changes since it was closed? I only ask because `@types/underscore` v1.9.x is still _ridiculously broken_ for chaining (updating from 1.8.x to 1.9.x breaks _a lot_ in our codebase from chaining alone) and it looks like you guys did a great job here. It would be awesome if we could get these changes in. I can help out, too, if needed!

**Comment by jacobbaskin** [comment]
Why don't you take over my branch and make a new PR? I just got swamped and
didn't respond to comments in time :-)

**Comment by jgonggrijp** [comment]
@jacobbaskin Thanks for updating us on this.

@kjleitz Let's finish this PR as @jacobbaskin invited us to do. You can do it if you want. I can also do it. Please let me know what you prefer.
